components/brave_new_tab_ui/components/default/braveToday/customize/useSearch.ts

```diff
@@ -0,0 +1,119 @@
+      .filter(p => p.publisherName.toLowerCase().includes(lowerQuery) ||
+        p.categoryName.toLocaleLowerCase().includes(lowerQuery) ||
+        p.siteUrl?.url?.toLocaleLowerCase().includes(query) ||
+        p.feedSource?.url?.toLocaleLowerCase().includes(query))
+    const results = { publishers, direct: [] as FeedSearchResultItem[] }
+    for (const result of directResults) {
+      const publisherMatch = publishers.find(p => p.feedSource.url === result.feedUrl.url)
+      if (publisherMatch && !publishers.some(p => p.publisherId === publisherMatch.publisherId)) {
+        results.publishers.push(publisherMatch)
+      } else {
+        results.direct.push(result)
+      }
+    }
+    return {
+      publishers: results.publishers.map(p => p.publisherId),
+      direct: results.direct
+    }
+  }, [directResults, context.sortedPublishers, lowerQuery])
+
+  React.useEffect(() => {
+    // Reset state when query changes
+    setCanFetchUrl(false)
+    setLoading(false)
+    setDirectResults([])
+  }, [query])
+
+  // Contains a url as string if the query looks like a url
+  const feedUrlQuery = React.useMemo<string | null>(() => {
+    // Check if we have a url-able query
+    let feedUrlRaw = query
+    // User inputting a protocol is a signal that they want to look
+    // for a feed. Otherwise we use the '.' that a hostname would have.
+    // This at least makes it possible for the user to retrieve local
+    // network feeds.
+    let queryContainsExplicitProtocol = false
+    if (feedUrlRaw.includes('://')) {
+      queryContainsExplicitProtocol = true
+    } else {
+      // Default protocol that should catch most cases. Make sure
+      // we check for validity after adding this prefix, and
+      // not before.
+      feedUrlRaw = 'https://' + feedUrlRaw
+    }
+    let isValidFeedUrl = false
+    try {
+      const url = new URL(feedUrlRaw)
+      isValidFeedUrl = ['http:', 'https:'].includes(url.protocol) &&
+        (url.hostname.includes('.') || queryContainsExplicitProtocol)
+    } catch { }
+    if (!isValidFeedUrl) {
+      // Not a url-able query
+      return null
+    }
+    return feedUrlRaw
+  }, [query])
+
+  React.useEffect(() => {
+    // Do nothing if user hasn't asked to search, or input isn't
+    // a valid Url
+    if (!canFetchUrl || !feedUrlQuery) {
+      console.debug('News: query changed but not searching', { canFetchUrl, feedUrlQuery })
+      return
+    }
+
+    setLoading(true)
+    let cancelled = false
+
+    api.controller.findFeeds({ url: feedUrlQuery.toString() }).then(({ results }) => {
+      console.debug('News: received feed results', { results, cancelled })
+      if (cancelled) return
+
+      setLoading(false)
+      setDirectResults(results)
+    })
+    return () => {
+      cancelled = true
+    }
+  }, [canFetchUrl, feedUrlQuery])
+
+  return {
+    loading,
+    filteredSources,
+    filteredChannels,
+    canFetchUrl,
+    setCanFetchUrl,
```

Comment:
Maybe we could actually hand out a `useCallback`'d function which directly does the search, rather than this implicit relationship with `canFetchUrl`

```ts
return {
    // other stuff
    canFetchUrl: !!feedUrlQuery,
    findDirectFeeds: () => api.controller.findFeeds(.....
}
```

Reply:
It's a good suggestion, probably more simpler. The refactor away from useEffect would require figuring out cancellation (probably using a Ref?), so I'd like to do as a follow-up. I don't think this is causing any issues aside from just an opinion on simplicity?

Reply:
Happy to leave this as a defer :smile